test/params_api_test.c

```diff
@@ -0,0 +1,362 @@
+    size_t bnsize;
+    BIGNUM *b = NULL, *c = NULL;
+    OSSL_PARAM param = OSSL_PARAM_DEFN("bn", OSSL_PARAM_UNSIGNED_INTEGER,
+                                       NULL, 0, NULL);
+    int ret = 0;
+
+    param.buffer = bnbuf;
+    param.buffer_size = len;
+    param.return_size = &bnsize;
+
+    copy_be_to_native(buf, raw_values[n].value, len);
+    swap_copy(le, raw_values[n].value, len);
+    if (!TEST_ptr(b = BN_bin2bn(raw_values[n].value, (int)len, NULL)))
+        goto err;
+
+    if (!TEST_true(OSSL_PARAM_set_BN(&param, b))
+        || !TEST_mem_eq(bnbuf, bnsize, buf, bnsize)
+        || !TEST_true(OSSL_PARAM_get_BN(&param, &c))
+        || !TEST_BN_eq(b, c))
+        goto err;
+
+    ret = 1;
+err:
+    BN_free(b);
+    BN_free(c);
+    return ret;
+}
+
+static int test_param_real(void)
+{
+    double p;
+    OSSL_PARAM param = OSSL_PARAM_double("r", NULL);
+
+    param.buffer = &p;
+    return TEST_true(OSSL_PARAM_set_double(&param, 3.14159))
+           && TEST_double_eq(p, 3.14159);
+}
+
+static int test_param_construct(void)
+{
+    static const char *number_names[] = {
+        "int", "uint", "long", "ulong", "int32", "uint32",
+        "int64", "uint64", "size_t"
+    };
+    static const unsigned char bn_val[16] = {
+        0xac, 0x75, 0x22, 0x7d, 0x81, 0x06, 0x7a, 0x23,
+        0xa6, 0xed, 0x87, 0xc7, 0xab, 0xf4, 0x73, 0x22
+    };
+    OSSL_PARAM params[20];
+    char buf[100], buf2[100], *bufp, *bufp2;
+    unsigned char ubuf[100];
+    void *vp, *vp2;
+    const OSSL_PARAM *p;
+    static const OSSL_PARAM pend = OSSL_PARAM_END;
+    int i, n = 0, ret = 0;
+    unsigned int u;
+    long int l;
+    unsigned long int ul;
+    int32_t i32;
+    uint32_t u32;
+    int64_t i64;
+    uint64_t u64;
+    size_t j, k, s, sz;
+    double d, d2;
+    BIGNUM *bn = NULL, *bn2 = NULL;
+
+    params[n++] = OSSL_PARAM_construct_int("int", &i, &sz);
+    params[n++] = OSSL_PARAM_construct_uint("uint", &u, &sz);
+    params[n++] = OSSL_PARAM_construct_long("long", &l, &sz);
+    params[n++] = OSSL_PARAM_construct_ulong("ulong", &ul, &sz);
+    params[n++] = OSSL_PARAM_construct_int32("int32", &i32, &sz);
+    params[n++] = OSSL_PARAM_construct_int64("int64", &i64, &sz);
+    params[n++] = OSSL_PARAM_construct_uint32("uint32", &u32, &sz);
+    params[n++] = OSSL_PARAM_construct_uint64("uint64", &u64, &sz);
+    params[n++] = OSSL_PARAM_construct_size_t("size_t", &s, &sz);
+    params[n++] = OSSL_PARAM_construct_double("double", &d, &sz);
+    params[n++] = OSSL_PARAM_construct_BN("bignum", ubuf, sizeof(ubuf), &sz);
+    params[n++] = OSSL_PARAM_construct_utf8_string("utf8str", buf, sizeof(buf),
+                                                   &sz);
+    params[n++] = OSSL_PARAM_construct_octet_string("octstr", buf, sizeof(buf),
+                                                    &sz);
+    params[n++] = OSSL_PARAM_construct_utf8_ptr("utf8ptr", &bufp, &sz);
+    params[n++] = OSSL_PARAM_construct_octet_ptr("octptr", &vp, &sz);
```

Comment:
Same parameter name again? Doesn't that defeat the purpose? `OSSL_PARAM_locate` will have a hard time finding these two

Reply:
Oh wait, never mind, I read the param names wrong